docs(solidity to MUD dictionary): first version (#2987)
latticexyz/mud: docs/pages/dictionary.mdx

import { Callout } from "nextra/components";

# Solidity to MUD Dictionary

To properly use MUD you need to do some things differently than the way you do them normally in Solidity.

## State variables

> A `System` should be stateless, for several reasons:

Reviewer: ```suggestion
A `System` contract should not store any internal state, but only interact with `World` state via `Tables`.
```
Making it more explicit that systems can work with state, just not internal state

- By concentrating all the state information in a single contract (the `World`) we make it easier to apply [access control](/world/namespaces-access-control), [execute hooks when data changes](/store/store-hooks), etc.
- It is a lot easier to [upgrade functionality](/cli/deploy#upgrading-a-world) when you do not need to deal with storage because the contract does not have a state.
- The same `System` can be used by multiple `World`s.

> The functionality of [state variables](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html) is implemented using [tables](/store/tables).

Reviewer: ```suggestion
Instead of using [state variables](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html) you can achieve the same functionality with [tables](/store/tables).
```

<details>

<summary>Single key mapping</summary>

```solidity filename="Solidity" {2}
contract Example {
  mapping(address => uint) public balances;
}
```

```typescript filename="MUD" {6-12}
import { defineWorld } from "@latticexyz/world";

export default defineWorld({
  namespace: "app",
  tables: {
    Balances: {
      schema: {
        owner: "address",
        balance: "uint256",
      },
      key: ["owner"],
    },
  },
});
```

</details>

<details>

<summary>Multiple key mapping</summary>

```solidity filename="Solidity" {2}
contract Example {
  mapping(address => mapping(address => uint256)) public allowances;
}
```

```typescript filename="MUD" {6-13}
import { defineWorld } from "@latticexyz/world";

export default defineWorld({
  namespace: "app",
  tables: {
    Allowances: {
      schema: {
        owner: "address",
        spender: "address",
        balance: "uint256",
      },
      key: ["owner", "spender"],
    },
  },
});
```

</details>

<Callout type="info" emoji="📝">
  MUD key fields have to be fixed-length.
  This means that some mappings, for example `mapping(string => uint) balances;`, cannot be directly translated to a MUD table.

> An easy workaround is to use the hash of the string as the key instead of the string itself.

Reviewer: ```suggestion
An easy workaround is to use the hash of the string as the key instead of the string itself (which is exactly what happens under the hood of vanilla Solidity with the keys of a mapping).
```

</Callout>

### Arrays

There are two ways in which MUD supports arrays.

1. Arrays of fixed-length value types, such as `uint24[]` or `address[]`, [are supported as table fields](https://github.com/latticexyz/mud/blob/main/packages/schema-type/src/solidity/SchemaType.sol#L108-L205).

   <details>

   <summary>Example</summary>

   ```solidity filename="Solidity" {2}
   contract Example {
     bool[] listOfBooleans;
   }
   ```

   ```typescript filename="MUD" {6-11}
   import { defineWorld } from "@latticexyz/world";

   export default defineWorld({
     namespace: "app",
     tables: {
       ListOfBooleans: {
         schema: {
           values: "bool[]",
         },
         key: [],
       },
     },
   });
   ```

   </details>

1. An array is a mapping between an unsigned integer and the value type of the array.
   We can use a table with `index` as the key. That way, we can get around the limitation that arrays of variable-length types such as `string` or `bytes` are not supported.

    <details>

>    {" "}

Reviewer: ```suggestion
```

Author: No. It doesn't change anything real, but I can't pass prettier without it.

   <summary>Example</summary>

   ```solidity filename="Solidity" {2}
   contract Example {
     string[] listOfNames;
   }
   ```

   ```typescript filename="MUD" {6-12}
   import { defineWorld } from "@latticexyz/world";

   export default defineWorld({
     namespace: "app",
     tables: {
       ListOfNames: {
         schema: {
           index: "uint256",
           name: "string",
         },
         key: ["index"],
       },
     },
   });
   ```

    </details>

{

> // ## Events

Reviewer: Why is this part commented out?

Author: Because I didn't manage to get it working on React with Zustand yet. So I decided it would be better to publish what I have and wait until we have stash to explain events.
